3rdparty/crypto/span.h

```diff
@@ -0,0 +1,378 @@
+// Copyright (c) 2018-2021 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#ifndef BITCOIN_SPAN_H
```

Comment:
I'd rather not bring in this custom implementation of span if we can avoid it (instead we should use `std::span` once C++20 is widely available).

Can we convert `Write` and `Finalize` to use the same interface as the existing `CSHA256` class? That way we don't need this header.

Reply:
cppreference [suggests](https://en.cppreference.com/w/cpp/compiler_support/20) that C++20 (and particularly, `std::span`) is rather widely-supported at this point. Should we just assume it's fair-game for us to jump to C++20 now?

Reply:
There are still several languages and especially standard library features missing from GCC/libstdc++ and Clang/libc++ so I don't think it's the right time to upgrade from C++17 just yet. Either way, this is not the appropriate PR to make such a change.

Reply:
action item: make SHA3 interface closer to SH256